add reusable helpers recipe and implement missing expression features (#1085)
kysely-org/kysely: site/docs/recipes/0001-reusable-helpers.md

# Reusable helpers

:::info
[Here's](https://kyse.link/cPY8e) a playground link containing all the code in this recipe.
:::

Let's say you want to write the following query:

```sql
SELECT id, first_name
FROM person
WHERE upper(last_name) = $1
```

Kysely doesn't have a built-in `upper` function but there are at least three ways you could write this:

```ts
const lastName = 'STALLONE'

const persons = await db
  .selectFrom('person')
  .select(['id', 'first_name'])

>   // 1. `sql` template tag. This is the least type-safe option

Reviewer: ```suggestion
  // 1. `sql` template tag. This is the least type-safe option. You're providing the column name without any type-checking.
```

Author: Good additions! I'll add these on separate comment lines though to keep the code more readable on mobile.

  .where(
    sql<string>`upper(last_name)`, '=', lastName
  )

>   // 2. `sql` template tag with `ref` for type-safe references

Reviewer: ```suggestion
  // 2. `sql` template tag with `ref`. Anything passed to `ref` gets checked against the accumulated query context, which uses your `Database` interface as source of truth.
```

  .where(({ eb, ref }) => eb(
    sql<string>`upper(${ref('last_name')})`, '=', lastName
  ))

>   // 3. The `fn` function

Reviewer: ```suggestion
  // 3. The `fn` function helps you avoid missing parentheses errors and uses refs as 1st class arguments.
```

  .where(({ eb, fn }) => eb(
    fn<string>('upper', ['last_name']), '=', lastName
  ))
  .execute()
```

> but each option could be more readable or type-safe.

Reviewer: ```suggestion
...but each option could be more readable or type-safe.
```

Here's an example of a helper function that uses the expression builder instead of raw SQL:

```ts
import { Expression, expressionBuilder } from 'kysely'

function idsOfPersonsThatHaveDogNamed(name: Expression<string>) {
  const eb = expressionBuilder<DB>()

  // A subquery that returns the identifiers of all persons
  // that have a dog named `name`.
  return eb
    .selectFrom('pet')
    .select('pet.owner_id')
    .where('pet.species', '=', 'dog')
    .where('pet.name', '=', name)
}
```

And here's how you could use it:

```ts
const dogName = 'Doggo'

const persons = await db
  .selectFrom('person')
  .selectAll('person')
  .where((eb) => eb(
    'person.id', 'in', idsOfPersonsThatHaveDogNamed(eb.val(dogName))
  ))
  .execute()
```

Note that `idsOfPersonsThatHaveDogNamed` doesn't execute a separate query but instead returns a subquery expression that's compiled as a part of the parent query:

```sql
select
  person.*
from
  person
where
  person.id in (
    select pet.owner_id
    from pet
    where pet.species = 'dog'
    and pet.name = ?
  )
```

In all our examples we've used the following syntax:

```ts
.where(eb => eb(left, operator, right))
```

> When the expression builder `eb` is used as a function, it creates a binary expression. All binary expression with a comparison operator are represented as a `Expression<SqlBool>`. You don't always need to return `eb(left, operator, right)` from the callback though. Since `Expressions` are composable and reusable, you can return any `Expression<SqlBool>`.

Reviewer: ```suggestion
When the expression builder `eb` is used as a function, it creates a binary expression. All binary expressions with a comparison operator are represented as `Expression<SqlBool>`. You don't always need to return `eb(left, operator, right)` from the callback though. Since `Expressions` are composable and reusable, you can return any `Expression<SqlBool>`.
```
